repos/system_upgrade/el7toel8/actors/checkpcidrivers/actor.py

```diff
@@ -0,0 +1,147 @@
+    )
+    unsupported_driver_names = ",\n".join(
+        driver
+        for driver in restricted_driver_names_on_host
+        if restricted_devices_drivers[driver]["supported_rhel8"] == 0
+    )
+    unsupported_pci_ids = ",\n".join(
+        pci_id
+        for pci_id in restricted_pci_ids_on_host
+        if restricted_devices_pcis[pci_id]["supported_rhel8"] == 0
+    )
+    title = "Detected PCI drivers that are restricted on RHEL8."
+    summary = (
+        "There are following drivers that are restricted on RHEL 8 system.\n\n"
+        "Devices which are unavailable on RHEL8:\n "
+        "{} (driver names),\n{} (pci ids)\n\n"
+        "Devices which are unsupported on RHEL8:\n "
+        "{} (driver names),\n{} (pci ids)\n\n"
+    ).format(
+        unavailable_driver_names,
+        unavailable_pci_ids,
+        unsupported_driver_names,
+        unsupported_pci_ids,
+    )
+    reports = [
+        reporting.Title(title),
+        reporting.Summary(summary),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Tags([reporting.Tags.KERNEL, reporting.Tags.DRIVERS]),
+    ]
+    if inhibit_upgrade:
+        reports += [reporting.Flags([reporting.Flags.INHIBITOR])]
+    return reports
+
+
+class CheckPCIDrivers(Actor):
+    """
+    Check if host PCI devices drivers has restrictions on target system.
+
+    If the driver is unavailable on a target system - then the upgrade will
+        be inhibited
+    If the driver is available, but not supported - then the upgrade will
+        continue, but the relevant report will be generated
+    If the driver is available and supported - then the upgrade will continue
+        normally
+    """
+
+    name = "check_pci_drivers"
+    consumes = (
+        PCIDevices,
+        RestrictedPCIDevices,
+    )
+    produces = (Report,)
+    tags = (ChecksPhaseTag, IPUWorkflowTag)
+
+    def process(self):
+        pci_devs = next(self.consume(PCIDevices))
+        try:
+            restricted_pci_devs = next(self.consume(RestrictedPCIDevices))
+        except StopIteration:
+            # when there is no restricted pci devices - it is fine and we
+            #   just do nothing
+            self.log.info(
+                "No data about restricted PCI devices. Skipping the actor."
+            )
+            return
+        else:
+            driver_names_on_host = set(dev.driver for dev in pci_devs.devices)
+            pci_ids_on_host = set(dev.pci_id for dev in pci_devs.devices)
+
+            restricted_devices_drivers = restricted_pci_devs.driver_names[
+                "devices"
+            ]
+            restricted_devices_pcis = restricted_pci_devs.pci_ids["devices"]
+
+            restricted_driver_names_on_host = driver_names_on_host & set(
+                restricted_devices_drivers
+            )
+            restricted_pci_ids_on_host = pci_ids_on_host & set(
+                restricted_devices_pcis
+            )
+
+            if any(
```

Comment:
Please at least not do the filtering in two places - I'd honestly prefer the restricted\unsupported filtering to be done here and passed to the render_report function (removing the L18-37).

Something like:
```
def _filter(col, key, match_against):
    return [obj for obj in col if match_against[obj][key] == 0]
```
```
restricted_devices_host = _filter(restricted_driver_names_on_host, "available_rhel8", restricted_devices_drivers)
restricted_pcis_host = _filter(restricted_pci_ids_on_host, "available_rhel8", restricted_devices_pcis)
unsupported_devices_host = _filter(restricted_driver_names_on_host, "supported_rhel8", restricted_devices_drivers)
unsupported_pcis_host = _filter(restricted_pci_ids_on_host, "supported_rhel8", restricted_devices_pcis)
```

Which will then be used in `process()` and passed to the `render_report(restricted_devices_host, ...)`

Reply:
thanks for your feedback. I specifically did `the filtering in two places`, because it works not like this actually.

The `filtering` inside the actor execution ends right after the first bad occurrence [1, 2].
And when something bad happen, we're proccessing the data in the way we need them for reporting.

Refs:
[1] The Python any (https://docs.python.org/3/library/functions.html#any) returns something on the first occurrence. 
[2] The same behavior is with `or`- interpreter tests the second statement only if first evaluated to False.
```python
In [1]: def a_lot_of_trues():
   ...:     import time
   ...:     time.sleep(10)
   ...:     return True
   ...: 

In [2]: True or a_lot_of_trues()
Out[2]: True    # no 10s wait
```

Reply:
That is imho unnecessary optimization. The code is really hard to follow the way it is. If others are ok with it - well, I will survive, but I can't put +1 on this unless the readability is improved, sorry.

Reply:
Golden rule is, that code has to be understood by others as ottherwise they cannot review it properly or they need to give it extra time. optimisation in cases where just couple of ms are saved, doesn't make any sense in this project, when the whole run takes tens minutes. In such cases, readability has higher prio the optimisation that nobody is able to realize. IOW, time spend by review is in such a case more valued than execution time.

As well, maintainability has to be counted too. When someone else will need to change something years later, we want to ensure the code is understood easily by everyone, so people will not rather create completly new code instead of updating it.

Good hint. When optimisation happens - or in general, when some code is not readable well by purpose - the comment should be placed in, to describe the code. if it's really necessary to keep the code as it is, why not to put the comment like
```
# the condition below is optimized version of code like:
# True in [obj for obj in col if match_against[obj][key] == 0]
```

Reply:
I will try to improve the readability

Reply:
@pirat89 @fernflower guys, I have added some comments to improve the readability. please check if it works for you now?